File: lab4-records/haskell/ttt.hs
module Main where

data Player 
    = Player_X | Player_O
    deriving (Eq,Show)

data Cell
    = X | O | B
    deriving (Eq)

instance (Show Cell) where
    show X = "X"
    show O = "O"
    show B = "."

data Board
    = Board [[Cell]] -- list of lists (ie rows) of cells
    deriving (Eq)

instance (Show Board) where
    show (Board rows) =
        -- show each row and put the rows on separate lines:
        unlines $ map showRow $ rows
            where
            showRow :: [Cell] -> String
            showRow cells =
                -- concatenate cells
                concat $ map show cells

-- start of code to focus on in Practical 4

data GamePos 
    = GamePos
      {
        pos_board :: Board,
        pos_turn :: Player,
        pos_value :: Float
      }
    deriving (Eq)

instance (Show GamePos) where
    show (GamePos board turn value) =
        unlines -- put the following strings on separate lines
        [show board, 
         "Player to take next move: " ++ show turn, 
         "Estimated advantage of player X against player O: " ++ show value]

gamePos1 =
    GamePos board Player_X 0
        where
        board = 
            Board
            [[X,X,O],
             [O,B,X],
             [O,B,B]]
  
gamePos1B =
    GamePos board Player_O 0
        where
        board = 
            Board
            [[X,X,O],
             [O,B,X],
             [O,B,B]]
  
{-
gamePos2 = 
    GamePosWin board Player_X
        where
        board = 
            Board
            [[X,X,O],
             [O,X,X],
             [O,O,X]]
-}

main =
    do
--    putStr "equalBoards (gamePos1, gamePos1B) = "
--    putStrLn $ show $ equalBoards (gamePos1, gamePos1B)
    putStrLn "gamePos1 = "
    putStrLn $ show gamePos1
--    putStrLn "gamePos2 = "
--    putStrLn $ show gamePos2

equalBoards pos1 pos2 = -- Task 4.2.(b): fix this function
    (pos_board pos1 == pos_board pos2)

File: lab4-records/java/ttt.java
public class ttt
{

    private static enum Player { X, O };
    
    private static enum CellValue { X, O, B };

    private static class Board
    {
        public Board(CellValue[][] rows)
        {
            this.rows = rows;
        }
        
        public CellValue[][] rows;
        
        public String toString()
        {
            String result = "";
            for ( CellValue[] row : rows)
            {
                for ( CellValue cell : row )
                {
                    result += cell;
                }
                result += "\n";
            }
            return result;
        }
    }
    
    
    private static class GamePos
    {
        public Board board;
        public Player turn;
        public float value;
        
        GamePos(Board board, Player turn, float value)
        {
            this.board = board;
            this.turn = turn;
            this.value = value;
        }
        
        public String toString()
        {
            String result = "";
            result += board;
            result += "Player to make the next move: ";
            result += turn + "\n";
            result += "Estimated advantage of Player X against Player O: ";
            result += value + "\n";
            
            return result;
        }
    }
    
    private static GamePos gamePos1 = 
        new GamePos
        (
            new Board
            (
                new CellValue[][]
                {{CellValue.X, CellValue.X, CellValue.O},
                 {CellValue.O, CellValue.B, CellValue.X},
                 {CellValue.O, CellValue.B, CellValue.B}}
            ),
            Player.X,
            0
        );
    
/*    
    private static GamePos gamePos2 = 
        new GamePosWon
        (
            new Board
            (
                new CellValue[][]
                {{CellValue.X, CellValue.X, CellValue.O},
                 {CellValue.O, CellValue.X, CellValue.X},
                 {CellValue.O, CellValue.O, CellValue.X}}
            ),
            Player.X
        );
*/
        
    public static void main(String[] args)
    {
        System.out.println(gamePos1);
//         System.out.println(gamePos2);
    }
}